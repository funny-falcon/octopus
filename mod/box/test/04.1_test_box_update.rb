#!/usr/bin/ruby1.9.1

$:.push 'test/lib'
require 'standalone_env'

class Env < StandAloneEnv
  def config
    super + <<EOD
object_space[0].enabled = 1
object_space[0].index[0].type = "HASH"
object_space[0].index[0].unique = 1
object_space[0].index[0].key_field[0].fieldno = 0
object_space[0].index[0].key_field[0].type = "NUM"
EOD
  end
end

Env.clean do |env|
  start

  c1 = env.connect
  c2 = env.connect
  c4 = env.connect

  wal_writer_pid = env.pid + 4 # hack!
  c1.insert [3, "baz"]
  puts "# wal_writer stop"
  Process.kill "STOP", wal_writer_pid
  t1 = Thread.new { log_try { c1.insert [1, "foo"] } }
  t2 = Thread.new { sleep 0.01; log_try { c2.insert [1, "foobar"] } }

  sleep 0.05
  c4.select 1

  puts "# wal_writer cont"
  Process.kill "CONT", wal_writer_pid
  t1.join
  t2.join
  c4.select 1
  c4.delete 1
  c4.select 1


  c1.insert [3, "bar"]
  puts "# wal_writer stop"
  Process.kill "STOP", wal_writer_pid
  t2 = Thread.new { log_try { c2.update_fields 3, [0, :set, 1] } }
  t1 = Thread.new { sleep 0.01; log_try { c1.insert [1, "foo"] } }
  sleep 0.05
  c4.select 1, 3
  Process.kill "CONT", wal_writer_pid
  puts "# wal_writer cont"
  t1.join
  t2.join
  c4.select 1
end
