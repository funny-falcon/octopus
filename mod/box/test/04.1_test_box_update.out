# box.insert([3, "baz"])
1

# wal_writer stop
# box.insert([1, "foo"])
# box.insert([1, "foobar"])
Failed with: {code: 0x401, message: 'object is locked'}
# box.select(1)
[]

# wal_writer cont
1

# box.select(1)
[["\x01\x00\x00\x00", "foo"]]

# box.delete(1)
1

# box.select(1)
[]

# box.insert([3, "bar"])
2

# wal_writer stop
# box.update_fields(3, [0, :set, 1])
# box.insert([1, "foo"])
Failed with: {code: 0x401, message: 'object is locked'}
# box.select(1, 3)
[["\x03\x00\x00\x00", "bar"]]

2

# wal_writer cont
# box.select(1)
[["\x01\x00\x00\x00", "bar"]]

